### DISCUSSION TITLE
[SOLVED] Cannot destroy dataset. Dataset is busy.

### DISCUSSION THREAD

**Discussion by ghost** [openzfs/zfs#13594] [General]
Hi there,
I've run into a problem that I cannot destroy a dataset, not after reboot and unmounting everything. The issue appeared after I tried to setup a ZFS pool for libvirt. I destroyed the pool in libvirt but cannot destroy the dataset on the disk. Any help is appreciated.
```
[user@computer]$ sudo zfs list -o name,mountpoint,mounted,type | grep libvirt
rpool_ab3df7/arch/DATA/default/var/lib/libvirt                                                                       /var/lib/libvirt          no       filesystem
rpool_ab3df7/arch/DATA/default/var/lib/libvirt/images                                                                /var/lib/libvirt/images   no       filesystem
rpool_ab3df7/arch/DATA/default/var/lib/libvirt/images/printvm                                                        -                         -        volume
[user@computer]$ sudo zfs destroy -f rpool_ab3df7/arch/DATA/default/var/lib/libvirt/images/printvm
cannot destroy 'rpool_ab3df7/arch/DATA/default/var/lib/libvirt/images/printvm': dataset is busy
```

**Comment by ghost** [comment]
OK, I was just able to resolve this. As you can see in the output I pasted above, the dataset I tried to destroy was a volume. This volume was mounted as a block device under /dev/zd0 (be aware the exact name can vary). I used gdisk to write a new partition header to the drive so that after a reboot no partitions from this volume will be mounted. I then rebooted and was able to destroy the dataset.

**Reply by dariusaurius** [reply]
That was the hint which helped me destroying the ZFS-Volume. However I have mounted the disk/volume to another Windows-VirtualMashine and used PowerShell to clear it:

`PS C:\> Clear-Disk -Number 1` (1 is most likely the second disk; check with "disk managment" the correct volume)

After that, destroying the RAW-Volume was no problem

**Reply by ElectricBill** [reply]
...aaaand you just helped me resolve the same problem here...inability to destroy zfs volume dataset even after reboot.

So I'll add a comment to return the favor.  I suspect that partition deletion and reboot is unnecessary,  One can probably use the kpartx command to delete the partition mappings instead.  I will try this myself soon because I will undoubtedly encounter the same condition again in my current testing.  But I figured I'd leave this note for others.  Thanks for your post - this was frustrating me too.

**Reply by no-usernames-left** [reply]
`sgdisk --zap-all /dev/zdX` sounds like it would do the trick nicely, no need for Windows (!).

**Comment by larz99** [comment]
Thanks for the note and follow-up comment!  I was pulling out my hair over this.

In my case, I had a Proxmox VM image with two partitions hanging out under /dev/zvol/rpool/data/vm-100-disk-1.  I believe I had deleted the VM before the image.

I used gdisk to delete the partitions, then expert mode, z-zap to wipe the GPT data.  I also deleted the old directories under /dev/zvol/rpool/data/.  I restarted Proxmox and removed the disk image references.

In my ignorance, I was probably a bit reckless in my approach, but I'm working on my own play system and trying to learn.

**Reply by no-usernames-left** [reply]
> I'm working on my own play system and trying to learn.

This is absolutely the best way to achieve that goal!
